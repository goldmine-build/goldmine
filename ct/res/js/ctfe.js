/**
 * Utility javascript functions used across the different CT FE pages.
 */
this.ctfe = this.ctfe || function() {
  "use strict";

  var ctfe = {};

  /**
   * Converts the timestamp used in CTFE DB into a user friendly string.
   **/
  ctfe.getFormattedTimestamp = function(timestamp) {
    if (timestamp == 0) {
      return "<pending>";
    }
    return ctfe.getTimestamp(timestamp).toLocaleString();
  }

  /**
   * Converts the timestamp used in CTFE DB into a Javascript timestamp.
   */
  ctfe.getTimestamp = function(timestamp) {
    if (timestamp == 0) {
      return timestamp;
    }
    var pattern = /(\d{4})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})/;
    return new Date(String(timestamp).replace(pattern,'$1-$2-$3T$4:$5:$6'));
  }

  /**
   * Convert from Javascript timestamp to one recognized by CTFE DB.
   */
  ctfe.getCtDbTimestamp = function(dbTimestamp) {
    var d = new Date(dbTimestamp);
    var timestamp = String(d.getUTCFullYear()) + sk.human.pad(d.getUTCMonth()+1, 2) +
                    sk.human.pad(d.getUTCDate(), 2) + sk.human.pad(d.getUTCHours(), 2) +
                    sk.human.pad(d.getUTCMinutes(), 2) + sk.human.pad(d.getUTCSeconds(), 2);
    return timestamp
  }

  /**
   * Get user friendly string for repeat after days.
   */
  ctfe.formatRepeatAfterDays = function(num) {
    if (num == 0) {
      return "N/A";
    } else if (num == 1) {
      return "Daily";
    } else if (num == 7) {
      return "Weekly";
    } else {
      return "Every " + num + " days";
    }
  }

  /**
   * Functions to work with information about page sets.
   */
  ctfe.pageSets = {};

  /**
   * Returns a Promise that resolves to an array of defined page sets.
   **/
  ctfe.pageSets.getPageSets = function() {
    return sk.post("/_/page_sets/")
        .then(JSON.parse);
  }

  /**
   * Returns an identifier for the given page set.
   **/
  ctfe.pageSets.getKey = function(pageSet) {
    return pageSet.key;
  }

  /**
   * Returns a short description for the given page set.
   **/
  ctfe.pageSets.getDescription = function(pageSet) {
    return pageSet.description;
  }

  /**
   * Functions to work with information about Chromium builds.
   */
  ctfe.chromiumBuild = {};

  /**
   * Returns a Promise that resolves to an array of completed builds.
   **/
  ctfe.chromiumBuild.getBuilds = function() {
    var queryParams = {
      "size": 20,
      "successful": true,
    }
    var queryStr = "?" + sk.query.fromObject(queryParams);
    return sk.post("/_/get_chromium_build_tasks" + queryStr)
        .then(JSON.parse)
        .then(function (json) {
          return json.data;
        });
  }

  /**
   * Returns a more human-readable GIT commit hash.
   */
  ctfe.chromiumBuild.shortHash = function(commitHash) {
    return commitHash.substr(0, 7);
  }

  /**
   * Returns a short description for the given build.
   **/
  ctfe.chromiumBuild.getDescription = function(build) {
    return ctfe.chromiumBuild.shortHash(build.ChromiumRev) + "-" +
        ctfe.chromiumBuild.shortHash(build.SkiaRev) + " (Chromium rev created on " +
        ctfe.getFormattedTimestamp(build.ChromiumRevTs) + ")";
  }

  /**
   * Returns a URL with details about the given Chromium commit hash.
   **/
  ctfe.chromiumBuild.chromiumCommitUrl = function(commitHash) {
    return "https://chromium.googlesource.com/chromium/src.git/+/" + commitHash;
  }

  /**
   * Returns a URL with details about the given Skia commit hash.
   **/
  ctfe.chromiumBuild.skiaCommitUrl = function(commitHash) {
    return "https://skia.googlesource.com/skia/+/" + commitHash;
  }

  /**
   * Functions to work with information about SKP repositories.
   */
  ctfe.skpRepositories = {};

  /**
   * Returns a Promise that resolves to an array of defined SKP repositories.
   **/
  ctfe.skpRepositories.getRepositories = function() {
    var queryParams = {
      "size": 20,
      "successful": true,
    }
    var queryStr = "?" + sk.query.fromObject(queryParams);
    return sk.post("/_/get_capture_skp_tasks" + queryStr)
        .then(JSON.parse)
        .then(function (json) {
          return json.data;
        });
  }

  /**
   * Returns a short description for the given SKP repository.
   **/
  ctfe.skpRepositories.getDescription = function(repository) {
    return repository.PageSets + " captured with " +
        ctfe.chromiumBuild.shortHash(repository.ChromiumRev) + "-" +
        ctfe.chromiumBuild.shortHash(repository.SkiaRev) + " (Desc: " +
        repository.Description + ")";
  }

  /**
   * Returns true and displays an error if user has more than 3 active tasks.
   */
  ctfe.moreThanThreeActiveTasks = function(sizeOfUserQueue) {
    if (sizeOfUserQueue > 3) {
        sk.errorMessage("You have " + sizeOfUserQueue + " currently running tasks. Please wait " +
                        "for them to complete before scheduling more CT tasks.");
    }
    return sizeOfUserQueue > 3;
  }

  /**
   * Returns a string that describes the specified CLs.
   **/
  ctfe.getDescriptionOfCls = function(chromiumClDesc, skiaClDesc, v8ClDesc, catapultClDesc) {
    if (!chromiumClDesc && !skiaClDesc && !v8ClDesc && !catapultClDesc) {
      return "";
    }
    var str = "Testing ";
    var prev = false;
    if (chromiumClDesc) {
      str += chromiumClDesc;
      prev = true;
    }
    if (skiaClDesc) {
      if (prev) {
        str += " and ";
      }
      str += skiaClDesc;
      prev = true;
    }
    if (v8ClDesc) {
      if (prev) {
        str += " and ";
      }
      str += v8ClDesc;
      prev = true;
    }
    if (catapultClDesc) {
      if (prev) {
        str += " and ";
      }
      str += catapultClDesc;
      prev = true;
    }
    return str;
  }

  return ctfe;
}();
