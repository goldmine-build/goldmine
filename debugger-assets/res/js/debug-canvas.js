// Abstracts out the server component in app.html.
//

var DebugCanvas = {
  // setBreakpoint runs the skp starting at opOffset until the pixel
  // located at (x,y) changes.
  //
  // Returns a promise that resolves with information on the changed pixel:
  //
  //   {
  //     "startColor":[255,79,54,200],
  //     "endColor":[255,144,206,1],
  //     "endOp":340
  //   }
  setBreakpoint: function(opOffset, x, y) {
    return sk.get("./break/" + opOffset + "/" + x + "/" + y).then(JSON.parse);
  },

  // getInfo returns a promise that resolves with information on
  // the view matrix and clip rect of the form:
  //
  //   {
  //     "ViewMatrix":[[1,0,0],[0,1,0],[0,0,1]],
  //     "ClipRect":[0,0,256,256],
  //   }
  getInfo: function(index) {
    if (index === undefined) {
      return sk.get("./info").then(JSON.parse);
    } else  {
      return sk.get("./info/" + index).then(JSON.parse);
    }
  },

  // getCmd returns a promise that resolves to a list of all the commands
  //   and the current state of the canvas:
  //
  //   {
  //     mode: "cpu",
  //     drawGpuOpBounds: false,
  //     colorMode: 0,
  //     version: 1,
  //     commands: [{command: "BeginDrawPicture", visible: true}, ...]
  //   }
  getCmd: function() {
      return sk.get("./cmd").then(JSON.parse);
  },

  // Sets the clip.
  //
  // b is a boolean.
  //
  // Return a promise that resolves when the change is complete.
  setClip: function(b) {
    var alpha = b ? 128 : 0;
    return sk.post("./clipAlpha/" + alpha);
  },

  // setTextBounds sets if text bounds are displayed.
  //
  // b is a boolean.
  //
  // Return a promise that resolves when the change is complete.
  setTextBounds: function(b) {
    var checked = b ? 1 : 0;
    return sk.post("./textBounds/" + checked);
  },

  // toggleOp turns the op at index off or on based on the boolean 'b'.
  //
  // Returns a promise that resolved when the change is complete.
  toggleOp: function(index, b) {
    var value = b ? "1" : "0";
    return sk.post("./cmd/" + index + "/" + value, "");
  },

  // setColorMode sets the color mode where mode is an integer, one of:
  //
  //   0 = Linear 32-bit
  //   1 = sRGB 32-bit
  //   2 = Linear half-precision float
  //
  // Returns a promise that resolved when the change is complete.
  setColorMode: function(mode) {
    return sk.post("./colorMode/" + mode, "");
  },

  // setGPU turns on or off the GPU based on the boolean b.
  //
  // Returns a promise that resolved when the change is complete.
  setGPU: function(b) {
    return sk.post("./enableGPU/" + (b ? 1 : 0), "");
  },

  // setGPUBounds turn on or off the display of the GPU bounds.
  //
  // Returns a promise that resolved when the change is complete.
  setGPUBounds: function (b) {
    return sk.post("./gpuOpBounds/" + (b ? 1 : 0), "");
  },
};
