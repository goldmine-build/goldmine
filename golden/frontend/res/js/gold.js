"use strict";

/* This defines the gold namespace which contains all JS code relevant to
   gold. Functions that are generic should move to common.js

   TODO(stephana): Move everythin the requires to be in sync with the
   backend to this file.
   */

var gold = gold || {};

(function(){
  // Constants for status values.
  gold.POSITIVE = 'positive',
  gold.NEGATIVE = 'negative';
  gold.UNTRIAGED = 'untriage';

  // Reference diffs.
  gold.REF_NEG   = 'neg';
  gold.REF_POS   = 'pos';
  gold.REF_TRACE = 'trace';

  // Metric values.
	gold.METRIC_COMBINED = 'combined';
	gold.METRIC_PERCENT  = 'percent';
	gold.METRIC_PIXEL    = 'pixel';
  gold.allMetrics = [
    gold.METRIC_COMBINED,
    gold.METRIC_PERCENT,
    gold.METRIC_PIXEL,
  ];

  // Operators to apply to images grouped by test.
  gold.GROUP_TEST_MAX_COUNT = 'count'    // Most often occuring digest.
  gold.groupTestOps = [
    gold.GROUP_TEST_MAX_COUNT,
  ];

  // ISSUE_TRACKER_URL is the url of the monorail issue tracker.
  var ISSUE_TRACKER_URL = 'https://bugs.chromium.org/p/skia/issues/';

  // Costants for sort order.
  gold.SORT_ASC = 'asc';
  gold.SORT_DESC = 'desc';
  gold.sortOptions = [
    gold.SORT_ASC,
    gold.SORT_DESC
  ];

  // Default values for the search controls.
  gold.defaultSearchState = {
    // The metric to use.
    metric: gold.METRIC_COMBINED,

    // Sort order.
    sort: gold.SORT_DESC,

    // Configs that need to match during comparisons.
    match: sk.app_config.defaultMatchFields,

    // Note: query is a URL encoded query over the test parameters
    // The fields of query are not fixed but change over time. This requires
    // to encode/decode a query in a separate step when encoding/decoding
    // this entire object.
    query:'',
    rquery: '',
    head: true,
    include: false,
    pos: false,
    neg: false,
    unt: true,
    blame: '',
    limit: 50,
    offset: 0,
    issue: '',
    patchsets: '',

    // Filter options.
    // Begin and end commits. Must be valid commits.
    fbegin: '',
    fend: '',

    // Select max RGBA difference.
    frgbamin: 0,

    // Select max RGBA difference.
    frgbamax: 255,

    // Select max difference.
    fdiffmax: -1,

    // Group by test and select a specific digest.
    fgrouptest: '',

    // Only include images that have a reference.
    fref: false,

    // master indicates whether to include digests that are also in master
    // when querying tryjob results.
    master: false,

    // use the new way of storing data belonging to changelists
    // https://bugs.chromium.org/p/skia/issues/detail?id=9340
    new_clstore: true,
  };

  // Default values for the search query of the by-blame-page.
  gold.defaultByBlameState = {
    query: '',
  };

  // Default values for pagination objects.
  gold.defaultPagination = {
    size: 50,
    offset: 0,
    total: 0
  };

  // Default values for pagination URL state.
  gold.defaultPaginationState = {
    size: gold.defaultPagination.size,
    offset: gold.defaultPagination.offset
  };

  // Table that maps reference point ids to readable titles.
  gold.diffTitles = {}
  gold.diffTitles[gold.REF_TRACE] = 'Trace previously';
  gold.diffTitles[gold.REF_POS] = 'Closest Positive';
  gold.diffTitles[gold.REF_NEG] = 'Closest Negative';

  // Return a title for the given reference point id.
  gold.getDiffTitle = function(diffType) {
    return gold.diffTitles[diffType] || diffType;
  };

  // Return the URL for the given digest.
  gold.imgHref = function(digest) {
    if (!digest) {
      return '';
    }

    return '/img/images/' + digest + '.png'
  };

  // Return the URL for the diff image between the two given digests.
  gold.diffImgHref = function(d1, d2) {
    if (!d1 || !d2) {
      return '';
    }

    return '/img/diffs/' + ((d1 < d2) ? (d1 + '-' + d2) : (d2 + '-' + d1)) + '.png'
  };

  // Returns the query string to pass to the diff page or to the diff endpoint.
  // Input is the name of the test and the two digests to compare.
  gold.diffQuery = function(test, left, right, issue) {
    const u = '?test=' + test + '&left=' + left + '&right=' + right;
    if (issue) {
      return u + '&issue=' +issue;
    }
    return u;
  };

  // Returns the query string to use for the detail page or the call to the
  // diff endpoint.
  gold.detailQuery = function(test, digest, issue) {
    const u = '?test=' + test + '&digest=' + digest;
    if (issue) {
      return u + '&issue=' +issue;
    }
    return u;
  };

  // stateFromQuery returns a state object based on the query portion of the URL.
  gold.stateFromQuery = function(defaultState) {
    var delta = sk.query.toObject(window.location.search.slice(1), defaultState);
    return sk.object.applyDelta(delta, defaultState);
  };

  // filterEmpty returns a copy of the object without fields where the value
  // is an empty string.
  gold.filterEmpty = function(obj) {
    var cpObj = {};
    for(var k in obj) {
      if (obj.hasOwnProperty(k) && (obj[k] !== '')) {
        cpObj[k] = obj[k];
      }
    }
    return cpObj;
  };

  // queryFromState returns a query string from the the given state object.
  gold.queryFromState = function(srcState) {
    var ret = sk.query.fromObject(gold.filterEmpty(srcState));
    if (ret === '') {
      return '';
    }
    return '?' + ret;
  };

  // updateParamsConditionally updates the given paramset 'params' with the
  // paramset 'updateParamSet' if the item is not present in the first.
  // If the 'force' flag it true it will always do the update.
  gold.updateParamsConditionally = function(params, updateParamSet, force) {
    for(var k in updateParamSet) {
      if (updateParamSet.hasOwnProperty(k) && (force || !params.hasOwnProperty(k))) {
        params[k] = updateParamSet[k];
      }
    }
    return sk.query.fromParamSet(params);
  }

  // loadWithActivity sends a GET request to the given url and uses the provide
  // acitivity element as an indicator. If the call succeeds it applies the
  // parsed result to 'target'.
  // If 'target' is a string it will call the 'set' function of the Polymer
  // element 'ele' with 'target', if 'target' is a function it will call it.
  gold.loadWithActivity = function(ele, url, activity, target) {
    activity.startSpinner('Loading...');
    sk.get(url).then(JSON.parse).then(function (json) {
      activity.stopSpinner();
      if (typeof(target) === 'function') {
        target(json);
      } else {
        ele.set(target, json);
      }
    }).catch(function(e) {
      activity.stopSpinner();
      sk.errorMessage(e);
    });
  },

  gold.issueURL = function(issueID) {
    return ISSUE_TRACKER_URL + '/detail?id=' + issueID;
  };

  // TriageQuery returns an object that can be sent as a query to the
  // backend to triage digests. The arguments can either be a 4-tuple:
  //    makeTriageQuery(testName, digests, status, issue)
  // or a pair:
  //    makeTriageQuery(arr, issue)
  // where 'arr' is an array of triples: <testName, digests, status>.
  // Note: 'digests' can either be a single string or an array.
  // 'issue' is the id of the code review issue for which we want to triage.
  // It has to be a positive integer (> 0) to be considered.
  gold.TriageQuery = function(triageList, issue) {
    if (arguments.length > 2) {
      triageList = [[arguments[0], arguments[1], arguments[2]]];
      issue = arguments[3];
    }

    var ret = {};
    triageList.forEach(function(t) {
      var test=t[0], digests=t[1], status=t[2];
      if (!Array.isArray(digests)) {
        digests = [digests];
      }

      var found = ret[test];
      if (!found) {
        ret[test] = {};
        found = ret[test];
      }

      for(var i=0; i < digests.length; i++) {
        found[digests[i]] = status;
      }
    });

    this.testDigestStatus = ret;
    this.setIssue(issue);
  };

  // setIssue is a setter for the issue of a TriageQuery
  // if "" or "0", will be assigned to the master branch.
  gold.TriageQuery.prototype.setIssue = function(issue) {
    this.issue = issue;
  };

  // flattenTriageQuery is the inverse operation of makeTriageQuery.
  // It returns an array of triples where each triple contains:
  //    [testName, digests, status]
  // testName and status are strings and digests is an array of strings.
  gold.flattenTriageQuery = function(q) {
    var ret = [];
    q = q.testDigestStatus
    for(var k in q) {
      if (q.hasOwnProperty(k)) {
        var statusMap = {};
        // iterat over the digests and group by status.
        for(var j in q[k]) {
          if (q[k].hasOwnProperty(j)) {
            var status = q[k][j];
            if (!statusMap[status]) {
              statusMap[status] = [];
            }
            statusMap[status].push(j);
          }
        }
        for(j in statusMap) {
          ret.push([k, statusMap[j], j]);
        }
      }
    }
    return ret;
  };

  // PageStateBehavior is a re-usable behavior what adds the _state and
  // _ctx (page.js context) variables to a Polymer element. All methods are
  // implemented as private since they should only be used within a
  // Polymer element.
  gold.PageStateBehavior = {
    properties: {
      _state: {
        type: Object,
        value: function() { return {}; }
      }
    },

    ready: function() {
      // Find the status element and listen to corpus changes.
      this.async(function() {
        this._statusElement = $$$('gold-status-sk');
        if (this._statusElement) {
          this.listen(this._statusElement, 'corpus-change', '_handleCorpusChange');
        }
      });
    },

    _handleCorpusChange: function(ev) {
      // Only change anything related to corpus if this element is the
      // part of the currently viewed page.
      if (Polymer.dom(this).parentNode.hasAttribute('activepage') && this._hasQuery) {
        if (this._corpusHome) {
          this._redirectHome();
          return;
        }
        var params = sk.query.toParamSet(this._state.query);
        params.source_type = [ev.detail];
        this._redirectToState({query: sk.query.fromParamSet(params)});
      }
    },

    // _setDefaultState sets the default state (usually reflected in the URL)
    // of the is document. 'corpusHome' is a boolean flag that indicates whether
    // a corpus change should redirect to the home page.
    _setDefaultState: function(defaultState, corpusHome) {
      this._defaultState = defaultState;
      this._corpusHome = corpusHome;
      this._hasQuery = defaultState.hasOwnProperty('query');
    },

    // _getDefaultStateWithCorpus returns the default search state of this
    // element (previously set via _setDefaultState) with the current corpus
    // injected.
    _getDefaultStateWithCorpus: function(state) {
        var ret = state || this._defaultState || {};
        if (this._statusElement && this._hasQuery) {
          ret = sk.object.shallowCopy(ret);
          ret.query = sk.query.fromParamSet({source_type: [this._statusElement.corpus]});
        }
        return ret;
    },

    // _initState initializes the '_state' and '_ctx' variables. ctx is the
    // context of the page.js route. It creates the value of the _state object
    // from the URL query string based on defaultState. It sets the URL to
    // the resulting the state.
    _initState: function(ctx, defaultState) {
      this._ctx = ctx;
      this._state = gold.stateFromQuery(defaultState);
      if (this._hasQuery) {
        this._syncCorpusQuery(defaultState.query);
      }
      this._setUrlFromState();
    },

    // _syncCorpusQuery synchronizes the the corpus value between the current
    // request (represented by this._state.query) with the corpus in status.
    // Effectively changing the corpus in status.
    _syncCorpusQuery: function(defaultQueryStr) {
      var defaultParams = sk.query.toParamSet(defaultQueryStr);
      var params = sk.query.toParamSet(this._state.query);
      this._state.query = gold.updateParamsConditionally(params, defaultParams, false);
      if (this._statusElement) {
        this._statusElement.setCorpus(params.source_type[0]);
      }
    },

    // _redirectToState updates the current state with 'updates'. After it
    // saves the current URL to history it redirects (via history.replaceState)
    // to newTargetPath, if provided, otherwise it will use the current path.
    _redirectToState: function(updates, newTargetPath) {
      // Save the current history entry before the redirect.
      this._ctx.pushState();
      page.redirect(this._getRedirectPath(updates, newTargetPath));
    },

    // Calculates a new path given the state update and an optional new target
    // path.
    _getRedirectPath: function(updates, newTargetPath) {
      var newState = this._addCorpus(sk.object.applyDelta(updates, this._state));
      var targetPath = newTargetPath ||  window.location.pathname;

      // TODO(stephana): Remove below if we can ever assign blame across corpora.
      // Account for the special case when the corpus changes and there is a
      // blame field. Then we want to go back to the by-blame-page.
      if ((this._statusElement) && !!newState.blame) {
        var newParams = sk.query.toParamSet(newState.query);
        var oldParams = sk.query.toParamSet(this._state.query);
        if (newParams.source_type[0] !== oldParams.source_type[0]) {
          this._redirectHome();
        }
      }
      return targetPath + gold.queryFromState(newState);
    },

    // _getRedirectURL returns a new URL based on the current state and
    // target path.
    _getRedirectURL: function(updates, newTargetPath) {
      var path = this._getRedirectPath(updates, newTargetPath);
      var host = window.location.protocol + '//' + window.location.hostname;
      return host + ':' + window.location.port + path;
    },

    // _redirectHome unconditionally redirects to home.
    _redirectHome: function() {
      this._ctx.pushState();
      page.redirect('/' + gold.queryFromState(this._getDefaultStateWithCorpus()));
    },

    // _replaceState updates the current state with 'updates' and updates
    // the URL accordingly. No new page is loaded or reloaded.
    _replaceState: function(updates) {
      this._state = sk.object.applyDelta(updates, this._state);
      this._setUrlFromState();
    },

    // setUrlFromState simply replaces the query string of the current URL
    // with a query string that represents the current state.
    _setUrlFromState: function() {
      history.replaceState(this._ctx.state, this._ctx.title, window.location.pathname + gold.queryFromState(this._state));
    },

    // _addCorpus injects the corpus into the query string of a query object.
    _addCorpus: function(state) {
      var params = sk.query.toParamSet(state.query);
      if ((!params['source_type]']) && this._statusElement) {
        params['source_type'] = [this._statusElement.corpus];
        state.query = sk.query.fromParamSet(params);
      }
      return state;
    }
  };

})();
