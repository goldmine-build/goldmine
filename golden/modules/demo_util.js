// delay wraps the value toReturn in a Promise that will resolve
// after "delay" milliseconds.
export function delay(toReturn, delay=100) {
  return new Promise((resolve) => {
    setTimeout(resolve, delay);
  }).then(() => {
    return {
      status: 200,
      body: JSON.stringify(toReturn),
      headers: {'content-type':'application/json'},
    };
  });
}