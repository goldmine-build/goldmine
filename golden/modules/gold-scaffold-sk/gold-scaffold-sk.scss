@import '~elements-sk/colors';
@import '../colors';

gold-scaffold-sk {
  app-sk {
    --header-background-color: var(--gold);
  }

  * {
    font-family:Roboto,Helvetica,Arial,'Bitstream Vera Sans',sans-serif;
  }

  header {
    h1 {
      font-weight: bold;
      font-size: 1.5em;
      padding: 0 0.25em;
    }

    .spinner-spacer {
      min-width: 40px;
    }

    spinner-sk {
      width: 24px;
      height: 24px;
      margin: 0;
      border: 8px solid var(--light-orange);
      border-left: 8px solid var(--red);
    }
  }

  nav {
    display:flex;
    flex-direction: column;
    a {
      padding: 0.5em 0.75em;
      color: var(--black);
      fill: var(--black);
      text-decoration: none;
    }
    span {
      padding: 0 0.5em;
      vertical-align: middle;
    }
  }

  login-sk {
    margin-left: auto;
    margin-right: 1em;
    display: block;
    color: white;
  }
}
