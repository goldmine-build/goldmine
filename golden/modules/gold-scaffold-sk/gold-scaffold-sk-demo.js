import './index';
import { testOnlySetSettings } from '../settings';
import { $$ } from 'common-sk/modules/dom';

testOnlySetSettings({
  title: 'Skia Public',
});
$$('gold-scaffold-sk')._render(); // pick up title from settings.
