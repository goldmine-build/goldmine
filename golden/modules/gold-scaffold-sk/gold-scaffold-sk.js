/**
 * @module modules/gold-scaffold-sk
 * @description <h2><code>gold-scaffold-sk</code></h2>
 *
 * Contains the title bar, side bar, and error-toast for all the Gold pages. The rest of
 * every Gold page should be a child of this element.
 *
 * Has a spinner-sk that can be activated when it hears "begin-task" events and keeps
 * spinning until it hears an equal number of "end-task" events.
 *
 * The error-toast element responds to fetch-error events and normal error-sk events.
 *
 * @attr {string} app_title - The title to show in the page banner.
 *
 * @attr {boolean} testing_offline - If we should operate entirely in offline mode.
 */
import { define } from 'elements-sk/define';
import { errorMessage } from 'elements-sk/errorMessage';
import { html } from 'lit-html';
import { ElementSk } from '../../../infra-sk/modules/ElementSk';

import '../../../infra-sk/modules/app-sk';
import '../../../infra-sk/modules/login-sk';

import 'elements-sk/error-toast-sk';
import 'elements-sk/icon/find-in-page-icon-sk';
import 'elements-sk/icon/folder-icon-sk';
import 'elements-sk/icon/help-icon-sk';
import 'elements-sk/icon/home-icon-sk';
import 'elements-sk/icon/label-icon-sk';
import 'elements-sk/icon/laptop-chromebook-icon-sk';
import 'elements-sk/icon/list-icon-sk';
import 'elements-sk/icon/search-icon-sk';
import 'elements-sk/icon/sync-problem-icon-sk';
import 'elements-sk/icon/view-day-icon-sk';
import 'elements-sk/spinner-sk';

const template = (ele) => html`
<app-sk>
  <header>
    <h1>${ele.appTitle}</h1>
    <div class=spinner-spacer>
      <spinner-sk></spinner-sk>
    </div>
    <div class=spacer></div>
    <!-- TODO(kjlubick) last commit -->
    <login-sk ?testing_offline=${ele.testingOffline}></login-sk>
  </header>

  <aside>
    <nav>
      <a href="/" tab-index=0>
        <home-icon-sk></home-icon-sk><span>Home</span>
      </a>
      <a href="/" tab-index=0>
        <view-day-icon-sk></view-day-icon-sk><span>By Blame<span>
      </a>
      <a href="/list" tab-index=0>
        <list-icon-sk></list-icon-sk><span>By Test</span>
      </a>
      <a href="/changelists" tab-index=0>
        <laptop-chromebook-icon-sk></laptop-chromebook-icon-sk><span>By ChangeList</span>
      </a>
      <a href="/search" tab-index=0>
        <search-icon-sk></search-icon-sk><span>Search</span>
      </a>
      <a href="/ignores" tab-index=0>
        <label-icon-sk></label-icon-sk><span>Ignores</span>
      </a>
      <a href="/triagelog" tab-index=0>
        <find-in-page-icon-sk></find-in-page-icon-sk><span>Triage Log</span>
      </a>
      <a href="/failures" tab-index=0>
        <sync-problem-icon-sk></sync-problem-icon-sk><span>Failures</span>
      </a>
      <a href="/help" tab-index=0>
        <help-icon-sk></help-icon-sk><span>Help</span>
      </a>
      <a href="https://github.com/google/skia-buildbot/tree/master/golden" tab-index=0>
        <folder-icon-sk></folder-icon-sk><span>Code</span>
      </a>
    </nav>
  </aside>

  <main></main>

  <footer><error-toast-sk></error-toast-sk></footer>
</app-sk>
`;

/**
 * Moves the elements from one NodeList to another NodeList.
 *
 * @param {NodeList} from - The list we are moving from.
 * @param {NodeList} to - The list we are moving to.
 */
function move(from, to) {
  Array.prototype.slice.call(from).forEach((ele) => to.appendChild(ele));
}

define('gold-scaffold-sk', class extends ElementSk {
  constructor() {
    super(template);
    this._main = null;
    this._busyTaskCount = 0;
    this._spinner = null;
  }

  connectedCallback() {
    super.connectedCallback();
    // Don't call more than once.
    if (this._main) {
      return;
    }
    this.addEventListener('begin-task', this._addBusyTask);
    this.addEventListener('end-task', this._finishedTask);
    this.addEventListener('fetch-error', this._fetchError);

    // We aren't using shadow dom so we need to manually move the children of
    // gold-scaffold-sk to be children of 'main'. We have to do this for the
    // existing elements and for all future mutations.

    // Create a temporary holding spot for elements we're moving.
    const div = document.createElement('div');
    move(this.children, div);

    // Now that we've moved all the old children out of the way we can render
    // the template.
    this._render();

    this._spinner = this.querySelector('header spinner-sk');

    // Move the old children back under main.
    this._main = this.querySelector('main');
    move(div.children, this._main);

    // Move all future children under main also.
    const observer = new MutationObserver((mutList) => {
      mutList.forEach((mut) => {
        move(mut.addedNodes, this._main);
      });
    });
    observer.observe(this, { childList: true });
  }

  disconnectedCallback() {
    super.disconnectedCallback();
    this.removeEventListener('begin-task', this._addBusyTask);
    this.removeEventListener('end-task', this._finishedTask);
    this.removeEventListener('fetch-error', this._fetchError);
  }

  /** @prop appTitle {string} Reflects the app_title attribute for ease of use. */
  get appTitle() { return this.getAttribute('app_title'); }

  set appTitle(val) { this.setAttribute('app_title', val); }

  /** @prop {boolean} busy Indicates if there any on-going tasks (e.g. RPCs).
   *                  This also mirrors the status of the embedded spinner-sk.
   *                  Read-only. */
  get busy() { return !!this._busyTaskCount; }

  /** @prop testingOffline {boolean} Reflects the testing_offline attribute for ease of use.
   */
  get testingOffline() { return this.hasAttribute('testing_offline'); }

  set testingOffline(val) {
    if (val) {
      this.setAttribute('testing_offline', '');
    } else {
      this.removeAttribute('testing_offline');
    }
  }

  /**
   * Indicate there are some number of tasks (e.g. RPCs) the app is waiting on
   * and should be in the "busy" state, if it isn't already.
   *
   */
  _addBusyTask() {
    this._busyTaskCount++;
    if (this._spinner && this._busyTaskCount > 0) {
      this._spinner.active = true;
    }
  }

  /**
   * Removes one task from the busy count. If there are no more tasks to wait
   * for, the app will leave the "busy" state and emit the "busy-end" event.
   *
   */
  _finishedTask() {
    this._busyTaskCount--;
    if (this._busyTaskCount <= 0) {
      this._busyTaskCount = 0;
      if (this._spinner) {
        this._spinner.active = false;
      }
      this.dispatchEvent(new CustomEvent('busy-end', { bubbles: true }));
    }
  }

  /** Handles a fetch error. Expects the detail of error to contain:
   *  error: the error given to fetch.
   *  loading: A string explaining what was being fetched.
   */
  _fetchError(e) {
    const loadingWhat = e.detail.loading;
    e = e.detail.error;
    if (e.name !== 'AbortError') {
      // We can ignore AbortError since they fire anytime an AbortController was canceled.
      // Chrome and Firefox report a DOMException in this case:
      // https://developer.mozilla.org/en-US/docs/Web/API/DOMException
      console.error(e);
      errorMessage(`Unexpected error loading ${loadingWhat}: ${e.message}`,
        5000);
    }
    this._finishedTask();
  }
});
