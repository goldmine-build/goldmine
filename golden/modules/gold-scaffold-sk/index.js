import './gold-scaffold-sk';
import './gold-scaffold-sk.scss';
