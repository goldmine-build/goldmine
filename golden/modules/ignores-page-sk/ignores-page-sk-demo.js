import './index.js'
import '../gold-scaffold-sk'

import { delay } from '../demo_util'
import { ignoreRules_10, fakeNow } from './test_data'

const fetchMock = require('fetch-mock');

Date.now = () => fakeNow;

fetchMock.get('/json/ignores?counts=1', delay(ignoreRules_10, 300));
fetchMock.post('glob:/json/ignores/del/*', delay({}, 600));
fetchMock.catch(404);
