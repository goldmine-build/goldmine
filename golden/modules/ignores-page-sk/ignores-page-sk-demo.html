<!DOCTYPE html>
<html>
<head>
    <title>ignore-page-sk demo</title>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>
<gold-scaffold-sk testing_offline>
    <ignores-page-sk page_size=10></ignores-page-sk>
</gold-scaffold-sk>
</body>
</html>
